include/kframework/lemmas/bytes-simplification.k

```diff
@@ -68,71 +68,168 @@ module BYTES-SIMPLIFICATION [symbolic]
 
     // #range
 
-    rule #range(_, _, K) => .Bytes requires K <=Int 0 [simplification]
+    rule [range-empty]:
+      #range(_:Bytes, S:Int, W:Int) => .Bytes
+      requires S <Int 0 orBool W <=Int 0
+      [simplification(30)]
+
+    rule [range-full]:
+      #range(B:Bytes, 0, W:Int) => B
+      requires lengthBytes(B) ==Int W
+      [simplification]
+
+    rule [range-overfull]:
+      #range(B:Bytes, 0, W:Int) => B +Bytes #buf(W -Int lengthBytes(B), 0)
+      requires lengthBytes(B) <Int W
+      [simplification]
+
+    rule [range-params-equal-ml]:
+      { #range (B1:Bytes, S1:Int, W1:Int) #Equals #range (B2:Bytes, S2:Int, W2:Int) } => #Top
+      requires B1 ==K B2 andBool S1 ==Int S2 andBool W1 ==Int W2
+      [simplification]
+
+    rule [range-included-in-cHead]:
+      #range(B1:Bytes +Bytes _:Bytes, S:Int, W:Int) => #range(B1, S, W)
+      requires S +Int W <=Int lengthBytes(B1)
+      [simplification]
+
+    rule [range-outside-cHead]:
+      #range(B1:Bytes +Bytes B2:Bytes, S:Int, W:Int) => #range(B2, S -Int lengthBytes(B1), W)
+      requires lengthBytes(B1) <=Int S
+      [simplification]
+
+    rule [range-includes-cHead]:
+      #range(B1:Bytes +Bytes B2:Bytes, 0, W:Int) => B1 +Bytes #range(B2, 0, W -Int lengthBytes(B1))
+      requires lengthBytes(B1) <=Int W
+      [simplification]
+
+    rule [range-inside-cHead-base-conc-left]:
+      #range(B:Bytes, S1:Int +Int S2:Int, W:Int) => #range(#range(B, S1, lengthBytes(B) -Int S1), S2, W)
+      requires 0 <=Int S1 andBool 0 <=Int S2 andBool 0 <Int S1 +Int S2 andBool S1 +Int S2 <=Int lengthBytes(B)
+      [concrete(B), concrete(S1), simplification]
+
+    rule [range-inside-cHead-base-conc-right]:
+      #range(B:Bytes, S1:Int +Int S2:Int, W:Int) => #range(#range(B, S2, lengthBytes(B) -Int S2), S1, W)
+      requires 0 <=Int S1 andBool 0 <=Int S2 andBool 0 <Int S1 +Int S2 andBool S1 +Int S2 <=Int lengthBytes(B)
+      [concrete(B), concrete(S2), simplification]
```

Comment:
We should already be moving all concrete terms in `+Int` to the right, so I don't think the first rule here is needed actually. Does it work if you remove it? https://github.com/runtimeverification/evm-semantics/blob/8a8dc22a2f56460339d6bc29fa90f6b7e58eb256/include/kframework/lemmas/int-simplification.k#L16

We could consider bumping the priority of the `+Int` and `+Bytes` re-associtation lemmas up to priority `40`.

Reply:
This works without change of priority.

Reply:
Yes, I assume it would. But these lemmas are redundant. If we're ni the situation `S1 +Int S2` with `concrete(S1), symbolic(S2)`, then the second rule won't fire, but we don't need to consider that case because the re-association lemma will flip around the `_+Int_`. So we really only need one of these lemmas, the second one (with `concrete(S2)`).

Suggesting to increase the priority of the reassociation lemma is just a side-thought, maybe it would be good to make sure that veerything as associated properly before applying other lemmas. Should not change the overall outcome.

Reply:
Yes yes, I agree, I meant that your suggestion works, we do need only one and the other passes as a claim. 

I'm happy to increase priority of associativity also.